Allow POMDP to state number of states

JupiLogy opened the pull request:
... instead of having to list names of states. This was also done for actions and observations.

JupiLogy committed:
Allow number of states/actions to be input, instead of enumerated by name

JupiLogy committed:
Changed pomdp.py to work for python 2.7

mbforbes commented:
Hi — Thank you for sending in this PR! Do you think you could provide a pomdp environment example that takes advantage of this format? It's been a while since I've looked at this code so it would be really helpful to me as a refresher 😅

JupiLogy commented:
Sure! [This maze](http://pomdp.org/examples/cheese.95.POMDP) has `states: 11`. You can find the [POMDP file spec](pomdp.org/code/pomdp-file-spec.html
), which describes this usage in the preamble. Also, there was a TODO in the code indicating it was intended to be implemented.

mbforbes commented on pomdp.py:
```diff
@@ -91,7 +92,7 @@ def belief_dump(self):
             self.belief = np.array([[b1], [b2]])
             best_action, reward = self.get_best_action()
-            print b1, b2, "\t", self.get_action_str(best_action)
+            print(b1, b2, "\t", self.get_action_str(best_action))
```
Let's keep the print statement instead of function (no parents) so that the whole file stays python 2.7 compatible. If we upgrade to the new py3 (and the print function), we can do it all at once :-)

mbforbes commented on pomdp.py:
```diff
@@ -163,19 +164,36 @@ def __get_value(self, i):
         self.states = line.split()[1:]
+        if len(self.states) == 1 and unicode(self.states[0], 'utf-8').isnumeric():
+            self.states = self.states[0]
+            self.no_states = int(self.states)
```
I'm not sure I understand why we need to introduce the `self.no_states` variable. Could we just keep the part you wrote that generates states based on the number: `self.states = [str(x) for x in range(1, self.no_states + 1)]`?

And as a followup, is there a reason to make these auto-generated lists 1-based instead of 0-based? (I.e., starting the `range` at 1?)

mbforbes commented on pomdp.py:
```diff
@@ -163,19 +164,36 @@ def __get_value(self, i):
-        # TODO support states as number
         line = self.contents[i]
         self.states = line.split()[1:]
+        if len(self.states) == 1 and unicode(self.states[0], 'utf-8').isnumeric():
```
Two thoughts:

1. Can we pull this check into a function like `is_numeric` so that we can reuse it for the different checks (and fix edge cases as they pop up)?

2. Instead of requiring the document to be `utf-8`, could we instead just call `int()` on the string, wrapped in a `try` block? I know that works on py3, but can't remember for py27.

mbforbes reviewed (changes requested):
Thanks for the link! Cool, this will be good to have. I've left some comments in the code review!

JupiLogy committed:
Fixing requested changes

Removed self.no_states variable, set state array to begin at 0, added is_numeric check

JupiLogy committed:
Shorten self.states/observations/transitions assignment when given by integer

JupiLogy commented:
I made all the changes!

JupiLogy committed:
Undoing shortening because lines too long

mbforbes commented on pomdp.py:
```diff
@@ -483,3 +499,14 @@ def get_best_action(self, belief):
         return (best_action, highest_expected_reward)
+
+
+def is_numeric(set):
```
nit: let's avoid `set` as it's a reserved word (it's a type). and i think this function operates on lists, so maybe `l` or `lst` or ... ?

mbforbes commented on pomdp.py:
```diff
@@ -241,14 +250,17 @@ def __get_transition(self, i):
         line = self.contents[i]
         pieces = [x for x in line.split() if (x.find(':') == -1)]
-        action = self.actions.index(pieces[0])
+        if pieces[0] == "*":
```
could you please add a comment for the new case this is handling?

mbforbes commented on pomdp.py:
```diff
@@ -205,7 +214,7 @@ def __get_transition(self, i):
             next_line = self.contents[i+1]
             probs = next_line.split()
-            assert len(probs) == len(self.states)
+            assert len(probs) == int(self.states)
```
let's keep as `len()` here :-)

mbforbes commented on pomdp.py:
```diff
@@ -163,19 +164,27 @@ def __get_value(self, i):
         self.states = line.split()[1:]
+        if is_numeric(self.states):
+            no_states = self.states[0]
+            self.states = [str(x) for x in range(int(no_states))]
```
For these three, cases, I'd either modify `no_` variables to be more clearly a number by changing the name and type, like:

```python
num_states = int(self.states[0])
```

or just remove it (since it's only used once) and instead have lines like:

```python
self.states = [str(x) for x in range(int(self.states[0]))]
```

your pick!

mbforbes reviewed (changes requested):
This is looking close! I'm so sorry for my slow reply over the holidays. Thank you so much for making these changes! Just a couple more small nits and I think this will be good to go!

JupiLogy committed:
Making more requested changes

Some naming changes mostly

JupiLogy committed:
Forgot to fix other cases

JupiLogy commented:
I think I've fixed it all now :) Let me know what you think!

mbforbes commented:
LGTM, thank you for this!
